crates/matrix-sdk/src/event_handler/maps.rs

```diff
@@ -27,25 +27,37 @@ pub(super) struct EventHandlerMaps {
     by_kind_type: BTreeMap<KindTypeWrap, Vec<EventHandlerWrapper>>,
     by_kind_roomid: BTreeMap<KindRoomId, Vec<EventHandlerWrapper>>,
     by_kind_type_roomid: BTreeMap<KindTypeRoomIdWrap, Vec<EventHandlerWrapper>>,
+    pub(super) notifications: Vec<EventHandlerWrapper<Notification>>,
 }
 
 impl EventHandlerMaps {
-    pub fn add(&mut self, handle: EventHandlerHandle, handler_fn: Box<EventHandlerFn>) {
+    pub fn add<E: ?Sized + 'static>(
+        &mut self,
+        handle: EventHandlerHandle,
+        handler_fn: Box<EventHandlerFn<E>>,
+    ) {
         let wrapper = EventHandlerWrapper { handler_id: handle.handler_id, handler_fn };
-
-        match Key::new(handle) {
-            Key::Kind(key) => {
-                self.by_kind.entry(key).or_default().push(wrapper);
-            }
-            Key::KindType(key) => {
-                self.by_kind_type.entry(key).or_default().push(wrapper);
-            }
-            Key::KindRoomId(key) => {
-                self.by_kind_roomid.entry(key).or_default().push(wrapper);
-            }
-            Key::KindTypeRoomId(key) => {
-                self.by_kind_type_roomid.entry(key).or_default().push(wrapper)
-            }
+        match try_downcast(wrapper) {
+            Ok(wrapper) => match Key::new(handle) {
```

Comment:
nit: maybe tweak the `wrapper` names so the generic type is mentioned? `Ok(wrapper)` -> `Ok(wrapper_raw_event)`